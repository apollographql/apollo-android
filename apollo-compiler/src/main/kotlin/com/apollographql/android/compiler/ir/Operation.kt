package com.apollographql.android.compiler.ir

import com.apollographql.android.compiler.SchemaTypeSpecBuilder
import com.apollographql.apollo.api.Operation
import com.squareup.javapoet.TypeSpec
import javax.lang.model.element.Modifier

data class Operation(
    val operationName: String,
    val operationType: String,
    val variables: List<Variable>,
    val source: String,
    val fields: List<Field>,
    val filePath: String,
    val fragmentsReferenced: List<String>
) : CodeGenerator {
  override fun toTypeSpec(context: CodeGenerationContext): TypeSpec =
      SchemaTypeSpecBuilder(DATA_TYPE_NAME, fields, emptyList(), emptyList(), context)
          .build(Modifier.PUBLIC, Modifier.STATIC)
          .toBuilder()
          .addSuperinterface(Operation.Data::class.java)
          .build()

  companion object {
    val DATA_TYPE_NAME = "Data"
  }
}
