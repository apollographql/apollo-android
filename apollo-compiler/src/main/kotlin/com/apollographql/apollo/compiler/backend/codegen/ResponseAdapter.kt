package com.apollographql.apollo.compiler.backend.codegen

import com.apollographql.apollo.api.CustomScalar
import com.apollographql.apollo.api.ResponseField
import com.apollographql.apollo.api.internal.ResponseAdapter
import com.apollographql.apollo.api.internal.ResponseReader
import com.apollographql.apollo.api.internal.ResponseWriter
import com.apollographql.apollo.compiler.applyIf
import com.apollographql.apollo.compiler.backend.ast.CodeGenerationAst
import com.apollographql.apollo.compiler.escapeKotlinReservedWord
import com.squareup.kotlinpoet.ClassName
import com.squareup.kotlinpoet.CodeBlock
import com.squareup.kotlinpoet.FunSpec
import com.squareup.kotlinpoet.KModifier
import com.squareup.kotlinpoet.ParameterSpec
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.PropertySpec
import com.squareup.kotlinpoet.TypeSpec
import com.squareup.kotlinpoet.asClassName
import com.squareup.kotlinpoet.asTypeName
import com.squareup.kotlinpoet.joinToCode

internal fun CodeGenerationAst.OperationType.responseAdapterTypeSpec(generateAsInternal: Boolean = false): TypeSpec {
  return TypeSpec.objectBuilder("${this.name.escapeKotlinReservedWord()}_ResponseAdapter")
      .applyIf(generateAsInternal) { addModifiers(KModifier.INTERNAL) }
      .addAnnotation(suppressWarningsAnnotation)
      .addSuperinterface(ResponseAdapter::class.asTypeName().parameterizedBy(this.dataType.typeRef.asTypeName()))
      .addProperty(responseFieldsPropertySpec(this.dataType.fields))
      .addFunction(
          FunSpec.builder("fromResponse")
              .addModifiers(KModifier.OVERRIDE)
              .returns(this.dataType.typeRef.asTypeName())
              .addParameter(ParameterSpec.builder("reader", ResponseReader::class).build())
              .addParameter(CodeGenerationAst.typenameField.asOptionalParameterSpec(withDefaultValue = false))
              .addCode("return·%T.fromResponse(reader, __typename)", this.dataType.typeRef.asAdapterTypeName())
              .build()
      )
      .addFunction(
          FunSpec.builder("toResponse")
              .addModifiers(KModifier.OVERRIDE)
              .addParameter(ParameterSpec(name = "writer", type = ResponseWriter::class.asTypeName()))
              .addParameter(ParameterSpec(name = "value", type = this.dataType.typeRef.asTypeName()))
              .addCode("%T.toResponse(writer, value)", this.dataType.typeRef.asAdapterTypeName())
              .build()
      )
      .addType(this.dataType.responseAdapterTypeSpec())
      .build()
}

internal fun CodeGenerationAst.FragmentType.responseAdapterTypeSpec(generateAsInternal: Boolean = false): TypeSpec {
  return TypeSpec.objectBuilder("${this.defaultImplementationType.name.escapeKotlinReservedWord()}_ResponseAdapter")
      .addSuperinterface(ResponseAdapter::class.asTypeName().parameterizedBy(this.defaultImplementationType.typeRef.asTypeName()))
      .applyIf(generateAsInternal) { addModifiers(KModifier.INTERNAL) }
      .applyIf(this.defaultImplementationType.fields.isNotEmpty()) {
        addProperty(responseFieldsPropertySpec(this@responseAdapterTypeSpec.defaultImplementationType.fields))
      }
      .addFunction(this.defaultImplementationType.readFromResponseFunSpec())
      .addFunction(this.defaultImplementationType.writeToResponseFunSpec())
      .addTypes(
          this.defaultImplementationType.nestedObjects.mapNotNull { nestedObject ->
            nestedObject
                .takeUnless { it.kind is CodeGenerationAst.ObjectType.Kind.Interface }
                ?.responseAdapterTypeSpec()
          }
      )
      .build()
}

private fun CodeGenerationAst.ObjectType.responseAdapterTypeSpec(): TypeSpec {
  return TypeSpec.objectBuilder(this.name)
      .addSuperinterface(ResponseAdapter::class.asTypeName().parameterizedBy(this.typeRef.asTypeName()))
      .applyIf(this.fields.isNotEmpty()) { addProperty(responseFieldsPropertySpec(this@responseAdapterTypeSpec.fields)) }
      .addFunction(readFromResponseFunSpec())
      .addFunction(writeToResponseFunSpec())
      .addTypes(
          this.nestedObjects
              .filter { it.kind !is CodeGenerationAst.ObjectType.Kind.Fragment || it.kind.possibleImplementations.isNotEmpty() }
              .mapNotNull { nestedObject ->
                nestedObject
                    .takeUnless { it.kind is CodeGenerationAst.ObjectType.Kind.Interface }
                    ?.responseAdapterTypeSpec()
              }
      )
      .build()
}

internal fun CodeGenerationAst.TypeRef.asAdapterTypeName(): ClassName {
  return if (enclosingType == null) {
    ClassName(packageName = "$packageName.adapter", "${this.name.escapeKotlinReservedWord()}_ResponseAdapter")
  } else {
    ClassName(packageName = "$packageName.adapter", enclosingType.asAdapterTypeName().simpleNames + this.name.escapeKotlinReservedWord())
  }
}

private fun responseFieldsPropertySpec(fields: List<CodeGenerationAst.Field>): PropertySpec {
  val initializer = CodeBlock.builder()
      .addStatement("arrayOf(")
      .indent()
      .add(fields.map { field -> field.responseFieldInitializerCode }.joinToCode(separator = ",\n"))
      .unindent()
      .addStatement("")
      .add(")")
      .build()
  return PropertySpec
      .builder(
          name = "RESPONSE_FIELDS",
          type = Array<ResponseField>::class.asClassName().parameterizedBy(ResponseField::class.asClassName()),
          modifiers = arrayOf(KModifier.PRIVATE)
      )
      .initializer(initializer)
      .build()
}

private val CodeGenerationAst.Field.responseFieldInitializerCode: CodeBlock
  get() {
    val factoryMethod = when (type) {
      is CodeGenerationAst.FieldType.Scalar -> when (type) {
        is CodeGenerationAst.FieldType.Scalar.ID -> "forString"
        is CodeGenerationAst.FieldType.Scalar.String -> "forString"
        is CodeGenerationAst.FieldType.Scalar.Int -> "forInt"
        is CodeGenerationAst.FieldType.Scalar.Boolean -> "forBoolean"
        is CodeGenerationAst.FieldType.Scalar.Float -> "forDouble"
        is CodeGenerationAst.FieldType.Scalar.Enum -> "forEnum"
        is CodeGenerationAst.FieldType.Scalar.Custom -> "forCustomScalar"
      }
      is CodeGenerationAst.FieldType.Object -> "forObject"
      is CodeGenerationAst.FieldType.Array -> "forList"
    }

    val builder = CodeBlock.builder().add("%T.%L", ResponseField::class, factoryMethod)
    when {
      type is CodeGenerationAst.FieldType.Scalar && type is CodeGenerationAst.FieldType.Scalar.Custom -> {
        builder.add("(%S,·%S,·%L,·%L,·%T,·%L)", responseName, schemaName, arguments.takeIf { it.isNotEmpty() }.toCode(), type.nullable,
            type.typeRef.asTypeName(), conditionsListCode(conditions))
      }
      else -> {
        builder.add("(%S,·%S,·%L,·%L,·%L)", responseName, schemaName, arguments.takeIf { it.isNotEmpty() }.toCode(), type.nullable,
            conditionsListCode(conditions))
      }
    }
    return builder.build()
  }

private fun conditionsListCode(conditions: Set<CodeGenerationAst.Field.Condition>): CodeBlock {
  return conditions
      .map { condition ->
        when (condition) {
          is CodeGenerationAst.Field.Condition.Directive -> CodeBlock.of("%T.booleanCondition(%S,·%L)",
              ResponseField.Condition::class, condition.variableName.escapeKotlinReservedWord(), condition.inverted)
        }
      }
      .joinToCode(separator = ",\n")
      .let {
        if (conditions.isEmpty()) {
          CodeBlock.of("null")
        } else {
          CodeBlock.builder()
              .add("listOf(\n")
              .indent().add(it).unindent()
              .add("\n)")
              .build()
        }
      }
}

private fun Any?.toCode(): CodeBlock {
  return when {
    this == null -> CodeBlock.of("null")
    this is Map<*, *> && this.isEmpty() -> CodeBlock.of("emptyMap<%T,·Any?>()", String::class.asTypeName())
    this is Map<*, *> -> CodeBlock.builder()
        .add("mapOf<%T,·Any?>(\n", String::class.asTypeName())
        .indent()
        .add(map { CodeBlock.of("%S to %L", it.key, it.value.toCode()) }.joinToCode(separator = ",\n"))
        .unindent()
        .add(")")
        .build()
    this is List<*> && this.isEmpty() -> CodeBlock.of("emptyList<Any?>()")
    this is List<*> -> CodeBlock.builder()
        .add("listOf<Any?>(\n")
        .indent()
        .add(map { it.toCode() }.joinToCode(separator = ",\n"))
        .unindent()
        .add(")")
        .build()
    this is String -> CodeBlock.of("%S", this)
    this is Number -> CodeBlock.of("%L", this)
    this is Boolean -> CodeBlock.of("%L", this)
    else -> throw IllegalStateException("Cannot generate code for $this")
  }
}
