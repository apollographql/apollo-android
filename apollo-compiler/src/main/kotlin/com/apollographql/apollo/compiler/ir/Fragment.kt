package com.apollographql.apollo.compiler.ir

import com.squareup.moshi.JsonClass

@JsonClass(generateAdapter = true)
data class Fragment(
    val fragmentName: String,
    val source: String,
    val description: String,
    val typeCondition: String,
    val possibleTypes: List<String>,
    val fields: List<Field>,
    val fragmentRefs: List<FragmentRef>,
    val inlineFragments: List<InlineFragment>,
    val filePath: String,
    val sourceLocation: SourceLocation
)