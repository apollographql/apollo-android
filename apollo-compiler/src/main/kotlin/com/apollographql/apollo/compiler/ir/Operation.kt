package com.apollographql.apollo.compiler.ir

import com.apollographql.apollo.api.Operation
import com.apollographql.apollo.api.internal.SimpleResponseWriter
import com.apollographql.apollo.compiler.Annotations
import com.apollographql.apollo.compiler.ClassNames
import com.apollographql.apollo.compiler.SchemaTypeSpecBuilder
import com.apollographql.apollo.compiler.withBuilder
import com.apollographql.apollo.response.ScalarTypeAdapters
import com.squareup.javapoet.MethodSpec
import com.squareup.javapoet.ParameterSpec
import com.squareup.javapoet.TypeSpec
import java.io.IOException
import javax.lang.model.element.Modifier

data class Operation(
    val operationName: String,
    val operationType: String,
    val variables: List<Variable>,
    val source: String,
    val sourceWithFragments: String,
    val fields: List<Field>,
    val filePath: String,
    val fragmentsReferenced: List<String>
) : CodeGenerator {

  override fun toTypeSpec(context: CodeGenerationContext, abstract: Boolean): TypeSpec =
      SchemaTypeSpecBuilder(
          typeName = DATA_TYPE_NAME,
          fields = fields,
          fragmentRefs = emptyList(),
          inlineFragments = emptyList(),
          context = context,
          abstract = abstract
      )
          .build(Modifier.PUBLIC, Modifier.STATIC)
          .toBuilder()
          .addSuperinterface(Operation.Data::class.java)
          .addMethod(toJsonWithDefaultScalarTypesMethodSpec())
          .addMethod(toJsonMethodSpec())
          .build()
          .let {
            if (context.generateModelBuilder) {
              it.withBuilder()
            } else {
              it
            }
          }

  fun normalizedOperationName(useSemanticNaming: Boolean): String = when (operationType) {
    TYPE_MUTATION -> normalizedOperationName(useSemanticNaming, "Mutation")
    TYPE_QUERY -> normalizedOperationName(useSemanticNaming, "Query")
    TYPE_SUBSCRIPTION -> normalizedOperationName(useSemanticNaming, "Subscription")
    else -> throw IllegalArgumentException("Unknown operation type $operationType")
  }

  private fun normalizedOperationName(useSemanticNaming: Boolean, operationNameSuffix: String): String {
    return if (useSemanticNaming && !operationName.endsWith(operationNameSuffix)) {
      operationName.capitalize() + operationNameSuffix
    } else {
      operationName.capitalize()
    }
  }

  fun isMutation() = operationType == TYPE_MUTATION

  fun isQuery() = operationType == TYPE_QUERY

  fun isSubscription() = operationType == TYPE_SUBSCRIPTION

  private fun toJsonWithDefaultScalarTypesMethodSpec(): MethodSpec {
    return MethodSpec
        .methodBuilder("toJson")
        .addAnnotation(Annotations.OVERRIDE)
        .addModifiers(Modifier.PUBLIC)
        .returns(String::class.java)
        .addParameter(ParameterSpec
            .builder(ClassNames.STRING, "indent")
            .addAnnotation(Annotations.NONNULL)
            .build()
        )
        .addStatement("return toJson(indent, \$T.DEFAULT)", ScalarTypeAdapters::class.java)
        .build()
  }

  private fun toJsonMethodSpec(): MethodSpec {
    return MethodSpec
        .methodBuilder("toJson")
        .addAnnotation(Annotations.OVERRIDE)
        .addModifiers(Modifier.PUBLIC)
        .returns(String::class.java)
        .addParameter(ParameterSpec
            .builder(ClassNames.STRING, "indent")
            .addAnnotation(Annotations.NONNULL)
            .build()
        )
        .addParameter(ParameterSpec
            .builder(ScalarTypeAdapters::class.java, "scalarTypeAdapters")
            .addAnnotation(Annotations.NONNULL)
            .build()
        )
        .addCode("""
            try {
              final ${'$'}T responseWriter = new ${'$'}T(scalarTypeAdapters);
              marshaller().marshal(responseWriter);
              return responseWriter.toJson(indent);
            } catch (${'$'}T e) {
              throw new ${'$'}T(e);
            }
           
        """.trimIndent(), SimpleResponseWriter::class.java, SimpleResponseWriter::class.java, IOException::class.java,
            RuntimeException::class.java)
        .build()
  }

  companion object {
    const val DATA_TYPE_NAME = "Data"
    const val TYPE_MUTATION = "mutation"
    const val TYPE_QUERY = "query"
    const val TYPE_SUBSCRIPTION = "subscription"
  }
}
