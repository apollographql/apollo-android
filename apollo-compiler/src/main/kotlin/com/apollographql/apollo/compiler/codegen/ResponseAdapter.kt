package com.apollographql.apollo.compiler.codegen

import com.apollographql.apollo.api.ResponseField
import com.apollographql.apollo.api.internal.ResponseAdapter
import com.apollographql.apollo.api.internal.ResponseReader
import com.apollographql.apollo.compiler.applyIf
import com.apollographql.apollo.compiler.ast.CodeGenerationAst
import com.squareup.kotlinpoet.ClassName
import com.squareup.kotlinpoet.CodeBlock
import com.squareup.kotlinpoet.FunSpec
import com.squareup.kotlinpoet.KModifier
import com.squareup.kotlinpoet.ParameterSpec
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import com.squareup.kotlinpoet.TypeName
import com.squareup.kotlinpoet.TypeSpec
import com.squareup.kotlinpoet.asTypeName
import com.squareup.kotlinpoet.joinToCode

internal fun CodeGenerationAst.OperationType.responseAdapterTypeSpec(targetPackage: String): TypeSpec {
  val dataImplementationType = checkNotNull(this.dataType.nestedTypes[this.dataType.rootType]) {
    "Failed to resolve operation root data type"
  }
  return TypeSpec.objectBuilder("${this.name}_ResponseAdapter")
      .addModifiers(KModifier.INTERNAL)
      .addAnnotation(suppressWarningsAnnotation)
      .addSuperinterface(ResponseAdapter::class.asTypeName().parameterizedBy(this.dataType.rootType.asTypeName()))
      .addProperty(responseFieldsPropertySpec(dataImplementationType.fields))
      .addFunction(
          dataImplementationType.readObjectFromResponseFunSpec(
              this.dataType.rootType
                  .copy(packageName = targetPackage)
                  .asTypeName()
          )
      )
      .addTypes(
          this.dataType
              .nestedTypes
              .minus(this.dataType.rootType)
              .filterNot { (_, type) -> type.abstract && type.kind !is CodeGenerationAst.ObjectType.Kind.Fragment }
              .map { (typeRef, type) ->
                type.responseAdapterTypeSpec(
                    typeRef
                        .copy(packageName = targetPackage)
                        .asTypeName()
                )
              }
      )
      .build()
}

internal fun CodeGenerationAst.FragmentType.responseAdapterTypeSpec(targetPackage: String): TypeSpec {
  val defaultImplementationType = checkNotNull(this.nestedTypes[this.defaultImplementation]) {
    "Failed to resolve operation root data type"
  }
  return TypeSpec.objectBuilder(this.rootType.asAdapterTypeName().simpleName)
      .addModifiers(KModifier.INTERNAL)
      .addAnnotation(suppressWarningsAnnotation)
      .addSuperinterface(ResponseAdapter::class.asTypeName().parameterizedBy(this.rootType.asTypeName()))
      .addProperty(responseFieldsPropertySpec(defaultImplementationType.fields))
      .addFunction(
          FunSpec.builder("fromResponse")
              .addModifiers(KModifier.OVERRIDE)
              .returns(this.rootType.asTypeName())
              .addParameter(ParameterSpec.builder("reader", ResponseReader::class).build())
              .addParameter(CodeGenerationAst.typenameField.asOptionalParameterSpec(withDefaultValue = false))
              .addStatement(
                  "return·%T.fromResponse(reader,·%L)",
                  ClassName(packageName = "", "${defaultImplementationType.name}_ResponseAdapter"),
                  CodeGenerationAst.typenameField.responseName
              )
              .build()
      )
      .addTypes(
          this.nestedTypes
              .minus(this.rootType)
              .filterNot { (_, type) -> type.abstract && type.kind !is CodeGenerationAst.ObjectType.Kind.Fragment }
              .map { (typeRef, type) ->
                type.responseAdapterTypeSpec(
                    typeRef
                        .copy(packageName = targetPackage)
                        .asTypeName()
                )
              }
      )
      .build()
}


private fun CodeGenerationAst.ObjectType.responseAdapterTypeSpec(objectTypeTypeName: TypeName): TypeSpec {
  return TypeSpec.objectBuilder("${this.name}_ResponseAdapter")
      .addSuperinterface(ResponseAdapter::class.asTypeName().parameterizedBy(objectTypeTypeName))
      .applyIf(fields.isNotEmpty()) { addProperty(responseFieldsPropertySpec(fields)) }
      .addFunction(
          when (this@responseAdapterTypeSpec.kind) {
            is CodeGenerationAst.ObjectType.Kind.Fragment -> readFragmentFromResponseFunSpec(objectTypeTypeName)
            is CodeGenerationAst.ObjectType.Kind.FragmentDelegate -> readFragmentDelegateFromResponseFunSpec(objectTypeTypeName)
            else -> readObjectFromResponseFunSpec(objectTypeTypeName)
          }
      )
      .build()
}

private fun CodeGenerationAst.ObjectType.readObjectFromResponseFunSpec(objectTypeTypeName: TypeName): FunSpec {
  val fieldVariablesCode = this.fields
      .map { field ->
        if (field.responseName == CodeGenerationAst.typenameField.responseName) {
          CodeBlock.of(
              "var·%L:·%T·=·%L",
              field.name,
              field.type.asTypeName().copy(nullable = true),
              CodeGenerationAst.typenameField.responseName
          )
        } else {
          CodeBlock.of(
              "var·%L:·%T·=·null",
              field.name,
              field.type.asTypeName().copy(nullable = true)
          )
        }
      }
      .joinToCode(separator = "\n", suffix = "\n")

  val selectFieldsCode = CodeBlock.builder()
      .beginControlFlow("while(true)")
      .beginControlFlow("when·(selectField(RESPONSE_FIELDS))")
      .add(
          this.fields.mapIndexed { fieldIndex, field ->
            CodeBlock.of(
                "%L·->·%L·=·%L",
                fieldIndex,
                field.name,
                field.type.nullable().fromResponseCode(field = "RESPONSE_FIELDS[$fieldIndex]")
            )
          }.joinToCode(separator = "\n", suffix = "\n")
      )
      .addStatement("else -> break")
      .endControlFlow()
      .endControlFlow()
      .build()

  val typeConstructorCode = CodeBlock.builder()
      .addStatement("%T(", objectTypeTypeName)
      .indent()
      .add(this.fields.map { field ->
        CodeBlock.of(
            "%L·=·%L%L",
            field.name,
            field.name,
            "!!".takeUnless { field.type.nullable } ?: ""
        )
      }.joinToCode(separator = ",\n", suffix = "\n"))
      .unindent()
      .addStatement(")")
      .build()

  return FunSpec.builder("fromResponse")
      .addModifiers(KModifier.OVERRIDE)
      .returns(objectTypeTypeName)
      .addParameter(ParameterSpec.builder("reader", ResponseReader::class).build())
      .addParameter(CodeGenerationAst.typenameField.asOptionalParameterSpec(withDefaultValue = false))
      .addCode(CodeBlock
          .builder()
          .beginControlFlow("return·reader.run")
          .add(fieldVariablesCode)
          .add(selectFieldsCode)
          .add(typeConstructorCode)
          .endControlFlow()
          .build()
      )
      .build()
}

private fun CodeGenerationAst.ObjectType.readFragmentFromResponseFunSpec(objectTypeTypeName: TypeName): FunSpec {
  val (defaultImplementation, possibleImplementations) = with(this.kind as CodeGenerationAst.ObjectType.Kind.Fragment) {
    defaultImplementation to possibleImplementations
  }
  return FunSpec.builder("fromResponse")
      .addModifiers(KModifier.OVERRIDE)
      .returns(objectTypeTypeName)
      .addParameter(ParameterSpec.builder("reader", ResponseReader::class).build())
      .addParameter(CodeGenerationAst.typenameField.asOptionalParameterSpec(withDefaultValue = false))
      .applyIf(possibleImplementations.isEmpty()) {
        addStatement(
            "return·%T.fromResponse(reader)",
            defaultImplementation.asAdapterTypeName()
        )
      }
      .applyIf(possibleImplementations.isNotEmpty()) {
        addStatement(
            "val·typename·=·%L·?:·reader.readString(RESPONSE_FIELDS[0])",
            CodeGenerationAst.typenameField.responseName
        )
        beginControlFlow("return·when(typename)")
        addCode(
            possibleImplementations
                .map { (typeCondition, type) ->
                  CodeBlock.of(
                      "%S·->·%T.fromResponse(reader,·typename)",
                      typeCondition,
                      type.asAdapterTypeName(),
                  )
                }
                .joinToCode(separator = "\n", suffix = "\n")
        )
        addStatement(
            "else·->·%T.fromResponse(reader, typename)",
            defaultImplementation.asAdapterTypeName()
        )
        endControlFlow()
      }
      .build()
}

private fun CodeGenerationAst.ObjectType.readFragmentDelegateFromResponseFunSpec(objectTypeTypeName: TypeName): FunSpec {
  val fragmentRef = (this.kind as CodeGenerationAst.ObjectType.Kind.FragmentDelegate).fragmentTypeRef
  return FunSpec.builder("fromResponse")
      .addModifiers(KModifier.OVERRIDE)
      .returns(objectTypeTypeName)
      .addParameter(ParameterSpec.builder("reader", ResponseReader::class).build())
      .addParameter(CodeGenerationAst.typenameField.asOptionalParameterSpec(withDefaultValue = false))
      .addStatement(
          "return·%T(%T.fromResponse(reader,·%L))",
          objectTypeTypeName,
          fragmentRef.asAdapterTypeName(),
          CodeGenerationAst.typenameField.responseName
      )
      .build()
}

internal fun CodeGenerationAst.TypeRef.asAdapterTypeName(): ClassName {
  return if (enclosingType == null) {
    ClassName(packageName = packageName, "${this.name}_ResponseAdapter")
  } else {
    ClassName(packageName = packageName, enclosingType.asAdapterTypeName().simpleName, "${this.name}_ResponseAdapter")
  }
}


private fun CodeGenerationAst.FieldType.fromResponseCode(field: String): CodeBlock {
  val notNullOperator = "!!".takeUnless { nullable } ?: ""
  return when (this) {
    is CodeGenerationAst.FieldType.Scalar -> when (this) {
      is CodeGenerationAst.FieldType.Scalar.ID -> if (field.isNotEmpty()) {
        CodeBlock.of("readCustomType<%T>(%L·as·%T)%L", ClassName.bestGuess(type), field, ResponseField.CustomTypeField::class,
            notNullOperator)
      } else {
        CodeBlock.of("readCustomType<%T>(%T)%L", ClassName.bestGuess(type), customEnumType.asTypeName(), notNullOperator)
      }
      is CodeGenerationAst.FieldType.Scalar.String -> CodeBlock.of("readString(%L)%L", field, notNullOperator)
      is CodeGenerationAst.FieldType.Scalar.Int -> CodeBlock.of("readInt(%L)%L", field, notNullOperator)
      is CodeGenerationAst.FieldType.Scalar.Boolean -> CodeBlock.of("readBoolean(%L)%L", field, notNullOperator)
      is CodeGenerationAst.FieldType.Scalar.Float -> CodeBlock.of("readDouble(%L)%L", field, notNullOperator)
      is CodeGenerationAst.FieldType.Scalar.Enum -> if (nullable) {
        CodeBlock.of("readString(%L)?.let·{·%T.safeValueOf(it)·}", field, typeRef.asTypeName().copy(nullable = false))
      } else {
        CodeBlock.of("%T.safeValueOf(readString(%L)!!)", typeRef.asTypeName().copy(nullable = false), field)
      }
      is CodeGenerationAst.FieldType.Scalar.Custom -> if (field.isNotEmpty()) {
        CodeBlock.of("readCustomType<%T>(%L·as·%T)%L", ClassName.bestGuess(type), field, ResponseField.CustomTypeField::class,
            notNullOperator)
      } else {
        CodeBlock.of(
            "readCustomType<%T>(%T)%L", ClassName.bestGuess(type), customEnumType.asTypeName().copy(nullable = false), notNullOperator
        )
      }
    }
    is CodeGenerationAst.FieldType.Object -> {
      val fieldCode = field.takeIf { it.isNotEmpty() }?.let { CodeBlock.of("(%L)", it) } ?: CodeBlock.of("")
      CodeBlock.builder()
          .addStatement("readObject<%T>%L·{·reader·->", typeRef.asTypeName(), fieldCode)
          .indent()
          .addStatement("%T.fromResponse(reader)", typeRef.asAdapterTypeName())
          .unindent()
          .add("}%L", notNullOperator)
          .build()
    }
    is CodeGenerationAst.FieldType.Array -> {
      CodeBlock.builder()
          .addStatement("readList<%T>(%L)·{·reader·->", rawType.asTypeName().copy(nullable = false), field)
          .indent()
          .add(rawType.readListItemCode())
          .unindent()
          .add("\n}%L", notNullOperator)
          .applyIf(!rawType.nullable) {
            if (nullable) {
              add("?.map·{ it!! }")
            } else {
              add(".map·{ it!! }")
            }
          }
          .build()
    }
  }
}

private fun CodeGenerationAst.FieldType.readListItemCode(): CodeBlock {
  return when (this) {
    is CodeGenerationAst.FieldType.Scalar -> when (this) {
      is CodeGenerationAst.FieldType.Scalar.ID -> CodeBlock.of(
          "reader.readCustomType<%T>(%T)", ClassName.bestGuess(type), customEnumType.asTypeName()
      )
      is CodeGenerationAst.FieldType.Scalar.String -> CodeBlock.of("reader.readString()")
      is CodeGenerationAst.FieldType.Scalar.Int -> CodeBlock.of("reader.readInt()")
      is CodeGenerationAst.FieldType.Scalar.Boolean -> CodeBlock.of("reader.readBoolean()")
      is CodeGenerationAst.FieldType.Scalar.Float -> CodeBlock.of("reader.readDouble()")
      is CodeGenerationAst.FieldType.Scalar.Enum -> CodeBlock.of(
          "%T.safeValueOf(reader.readString())", typeRef.asTypeName().copy(nullable = false)
      )
      is CodeGenerationAst.FieldType.Scalar.Custom -> CodeBlock.of(
          "reader.readCustomType<%T>(%T)", ClassName.bestGuess(type), customEnumType.asTypeName()
      )
    }
    is CodeGenerationAst.FieldType.Object -> {
      CodeBlock.builder()
          .addStatement("reader.readObject<%T>·{·reader·->", typeRef.asTypeName())
          .indent()
          .addStatement("%T.fromResponse(reader)", typeRef.asAdapterTypeName())
          .unindent()
          .add("}")
          .build()
    }
    is CodeGenerationAst.FieldType.Array -> {
      CodeBlock.builder()
          .addStatement("reader.readList<%T>·{·reader·->", rawType.asTypeName().copy(nullable = false))
          .indent()
          .add(rawType.readListItemCode())
          .unindent()
          .add("\n}")
          .applyIf(!rawType.nullable) { add(".map·{ it!! }") }
          .build()
    }
  }
}
