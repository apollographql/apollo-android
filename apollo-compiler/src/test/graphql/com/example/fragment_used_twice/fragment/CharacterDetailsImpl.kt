// AUTO-GENERATED FILE. DO NOT MODIFY.
//
// This class was automatically generated by Apollo GraphQL plugin from the GraphQL queries it found.
// It should not be modified by hand.
//
package com.example.fragment_used_twice.fragment

import com.apollographql.apollo.api.CustomScalarAdapters
import com.apollographql.apollo.api.Fragment
import com.apollographql.apollo.api.Operation
import com.apollographql.apollo.api.ResponseField
import com.apollographql.apollo.api.internal.ResponseAdapter
import com.example.fragment_used_twice.fragment.adapter.CharacterDetailsImpl_ResponseAdapter
import kotlin.Any
import kotlin.String
import kotlin.collections.List

class CharacterDetailsImpl : Fragment<CharacterDetailsImpl.Data> {
  override fun adapter(customScalarAdapters: CustomScalarAdapters): ResponseAdapter<Data> {
    val adapter = customScalarAdapters.getFragmentAdapter("CharacterDetailsImpl") {
      CharacterDetailsImpl_ResponseAdapter(customScalarAdapters)
    }
    return adapter
  }

  override fun responseFields(): List<ResponseField.FieldSet> = listOf(
    ResponseField.FieldSet(null, CharacterDetailsImpl_ResponseAdapter.RESPONSE_FIELDS)
  )
  override fun variables(): Operation.Variables = Operation.EMPTY_VARIABLES

  /**
   * A character from the Star Wars universe
   */
  data class Data(
    override val __typename: String = "Character",
    /**
     * The name of the character
     */
    override val name: String,
    /**
     * The date character was born.
     */
    override val birthDate: Any
  ) : CharacterDetails, Fragment.Data
}
