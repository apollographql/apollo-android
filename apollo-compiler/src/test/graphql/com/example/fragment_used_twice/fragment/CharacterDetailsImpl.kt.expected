// AUTO-GENERATED FILE. DO NOT MODIFY.
//
// This class was automatically generated by Apollo GraphQL plugin from the GraphQL queries it found.
// It should not be modified by hand.
//
package com.example.fragment_used_twice.fragment

import com.apollographql.apollo3.api.Fragment
import com.apollographql.apollo3.api.ResponseAdapterCache
import com.apollographql.apollo3.api.ResponseField
import com.apollographql.apollo3.api.internal.ResponseAdapter
import com.apollographql.apollo3.api.internal.json.JsonWriter
import com.example.fragment_used_twice.fragment.adapter.CharacterDetailsImpl_Adapter
import com.example.fragment_used_twice.fragment.adapter.CharacterDetailsImpl_ResponseAdapter
import kotlin.Any
import kotlin.String
import kotlin.collections.List

class CharacterDetailsImpl : Fragment<CharacterDetailsImpl.Data> {
  override fun adapter(responseAdapterCache: ResponseAdapterCache): ResponseAdapter<Data> {
    val adapter = responseAdapterCache.getAdapterFor(this::class) {
      CharacterDetailsImpl_ResponseAdapter(responseAdapterCache)
    }
    return adapter
  }

  override fun responseFields(): List<ResponseField.FieldSet> = listOf(
    ResponseField.FieldSet(null, CharacterDetailsImpl_ResponseAdapter.RESPONSE_FIELDS)
  )
  override fun serializeVariables(writer: JsonWriter, responseAdapterCache: ResponseAdapterCache) {
    responseAdapterCache.getVariablesAdapterFor(this::class) {
      CharacterDetailsImpl_Adapter(responseAdapterCache)
    }.toResponse(writer, this)
  }

  /**
   * A character from the Star Wars universe
   */
  data class Data(
    override val __typename: String = "Character",
    /**
     * The name of the character
     */
    override val name: String,
    /**
     * The date character was born.
     */
    override val birthDate: Any
  ) : CharacterDetails, Fragment.Data
}
